src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs

```diff
@@ -476,22 +475,68 @@ TypeSymbol BindRecursivePatternType(TypeSyntax typeSyntax, TypeSymbol inputType,
             {
-                deconstructionSubpatterns = BindDeconstructionPatternClause(node.DeconstructionPatternClause, declType, inputValEscape, diagnostics, out deconstructMethod, ref hasErrors);
+                DeconstructionPatternClauseSyntax deconstructClause = node.DeconstructionPatternClause;
+                Debug.Assert(deconstructClause != null);
+                deconstructMethod = null;
+                var patterns = ArrayBuilder<BoundSubpattern>.GetInstance(deconstructClause.Subpatterns.Count);
+                if (IsZeroElementTupleType(declType))
+                {
+                    // Work around https://github.com/dotnet/roslyn/issues/20648: The compiler's internal APIs such as `declType.IsTupleType`
+                    // do not correctly treat the non-generic struct `System.ValueTuple` as a tuple type.  We explicitly perform the tests
+                    // required to identify it.  When that bug is fixed we should be able to remove this if statement.
+                    BindZeroElementValueTupleSubpatterns(deconstructClause, declType, patterns, inputValEscape, ref hasErrors, diagnostics);
+                }
+                else if (declType.IsTupleType)
+                {
+                    // It is a tuple type. Work according to its elements
+                    BindValueTupleSubpatterns(deconstructClause, declType, inputValEscape, ref hasErrors, patterns, diagnostics);
+                }
+                else
+                {
+                    // It is not a tuple type. Seek an appropriate Deconstruct method.
+                    var inputPlaceholder = new BoundImplicitReceiver(deconstructClause, declType); // A fake receiver expression to permit us to reuse binding logic
+                    var deconstructDiagnostics = DiagnosticBag.GetInstance();
+                    BoundExpression deconstruct = MakeDeconstructInvocationExpression(
+                        deconstructClause.Subpatterns.Count, inputPlaceholder, deconstructClause, deconstructDiagnostics, outPlaceholders: out ImmutableArray<BoundDeconstructValuePlaceholder> outPlaceholders,
+                        out bool anyDeconstructCandidates);
+                    if (!anyDeconstructCandidates && ShouldUseITupleForRecursivePattern(node, declType, diagnostics, out var iTupleType, out var iTupleGetLength, out var iTupleGetItem))
+                    {
+                        // There was no Deconstruct, but the constraints for the use of ITuple are satisfied.
+                        // Use that and forget any errors from trying to bind Deconstruct.
+                        deconstructDiagnostics.Free();
+                        BindITupleSubpatterns(deconstructClause, patterns, diagnostics);
+                        deconstructionSubpatterns = patterns.ToImmutableAndFree();
+                        return new BoundITuplePattern(node, iTupleGetLength, iTupleGetItem, deconstructionSubpatterns, inputType, hasErrors);
```

Comment:
>return new BoundITuplePattern [](start = 24, length = 29)

Should `BindDeconstructSubpatterns()` be called? Might there be errors in the sub-patterns that are not related to the missing `Deconstruct`?